### ISSUE TITLE
Red speed of this script is far below Rebol

### ISSUE THREAD

**Issue by vdsfr** [red/red#1972]
```
 Red [ ]
; rebol[]
;
; subject: execution time Red
; Question: Why is Red so slow
;
; note: It's not about how to improve the code
;
;; execution times
;  rebol:     3   seconds
;  red:      13   seconds 
;  compiled: 10   seconds

p-mod: func [
  ;  title: "Compute a nonnegative remainder of A divided by B."
  ;  [catch]
    a [number!  ]
    b [number!   ] "Must be nonzero."
    /local r
][
    all [negative? r: a // b r: r + b]
   ; a: abs a
    a: load skip p-to-string a
    either all [a + r = (a + b) positive? r + r - b] [r - b] [r]
]

;--
p-to-integer: func [value][to integer! :value]
;--
p-to-string: func [value][to string! :value]
;--
 p-rejoin: func [
    block [block!] "Values to reduce and join"
   ][
    if empty? block: reduce block [return block]
    append either series? first block [copy first block] [
        form first block
    ] next block
]
;--
 add-digs: func [digs [block!]][
	 out-digs: copy []
	 
	 ; convert to integer
	 nbrs: ""
     foreach dig digs [  
          nbrs:  p-rejoin [nbrs dig]
      ]
      ; increment
      int-nbr: p-to-integer nbrs
      int-nbr: int-nbr + 1
      if (p-mod int-nbr 10) = 0 [int-nbr: int-nbr + 1]
      
      ; convert back to block
      foreach el p-to-string int-nbr [
          append out-digs p-to-integer ( el - 48)
      ]
    out-digs
]
digs: [ 1 2 3 4 5]
loop 87654 [
    digs: add-digs digs
]
print mold digs
 halt
```

**Comment by dockimbel** [comment]
First of all, I rewrote the title of your ticket as you are *wrongly* drawing a general conclusion from *one very specific* test. I warn people against micro-benchmarking especially because of that.

You Red code above is slower than Rebol because it relies on `load` which implementation in Red is currently done at high-level (for simplicity and flexibility), while Rebol's implementation is done at low-level using a very optimized algorithm. Remove that dependency, and you'll see your compiled Red code perform faster than Rebol.

We will provide in the future an alternative `load` implementation that will perform at the same speed as Rebol's one. In the meantime, don't use it in any intensive loading task (like loading millions of values) or in any benchmark.

**Comment by vdsfr** [comment]
commented out the line (see below)  to avoid 'load.
The compiled version execution time went from 13 to 7 seconds.
I suppose that there are one or more high-level implementations involved  that will be improved over time?
 
if (p-mod int-nbr 10) = 0 [int-nbr: int-nbr + 1]

**Comment by dockimbel** [comment]
`[to integer! :value]` is calling `load` internally.

`negative?`, `positive?`, `series?`, `first` are defined as mezzanine functions in Red, while they are natives in Rebol (difference in speed is big between mezz and natives).

**Comment by dockimbel** [comment]
The following code replaces the mezzanines with alternative native functions (though still not as good as making the mezzanines native):
```
 Red [ ]
 rebol[]
;
; subject: execution time Red
; Question: Why is Red so slow
;
; note: It's not about how to improve the code
;
;; execution times
;  rebol:     3   seconds
;  red:      13   seconds 
;  compiled: 10   seconds

to-integer: func [s][
	either string! = type? s [
		system/lexer/make-number s tail s integer!
	][
		0 + s				;-- char! case
	]
]

p-mod: func [
  ;  title: "Compute a nonnegative remainder of A divided by B."
  ;  [catch]
    a [number!  ]
    b [number!   ] "Must be nonzero."
    /local r
][
    all [(r: a // b) < 0 r: r + b]
    a: absolute a
    ;a: load skip p-to-string a
    either all [a + r = (a + b) r + r - b > 0] [r - b] [r]
]

;--
p-to-integer: func [value][to-integer :value]
;p-to-integer: func [value][to integer! :value]
;--
p-to-string: func [value][to string! :value]
;--
 p-rejoin: func [
    block [block!] "Values to reduce and join"
   ][
    if empty? block: reduce block [return block]
    append either series! = type? pick block 1 [copy pick block 1] [
        form pick block 1
    ] next block
]
;--
 add-digs: func [digs [block!]][
     out-digs: copy []

     ; convert to integer
     nbrs: ""
     foreach dig digs [  
          nbrs:  p-rejoin [nbrs dig]
      ]
      ; increment
      int-nbr: p-to-integer nbrs
      int-nbr: int-nbr + 1
      if (p-mod int-nbr 10) = 0 [int-nbr: int-nbr + 1]

      ; convert back to block
      foreach el p-to-string int-nbr [
          append out-digs p-to-integer ( el - 48)
      ]
    out-digs
]
digs: [ 1 2 3 4 5]
loop 200000 [
    digs: add-digs digs
]
print mold digs
 halt
```

**Comment by dockimbel** [comment]
I'm closing this ticket as it's anyway irrevelant for now. Performance tuning should be done in 0.9.x releases, while full optimization will be implemented in 2.0.

Also, before 1.0, we plan to provide an alternative code generator for Red that will offer some significant speed gains (though, that's all I can disclose for now).
